File: app/api/v1/endpoints/notifications.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from pydantic import BaseModel

from app.db.database import get_db
from app.models.notification import Notification, NotificationStatus, NotificationType, NotificationTemplateType
from app.models.user import User, UserRole
from app.api.v1.endpoints.auth import get_verified_user
from app.services.notification_service import notification_service

router = APIRouter()

class NotificationResponse(BaseModel):
    id: int
    user_id: int
    case_id: Optional[int]
    type: NotificationType
    template: NotificationTemplateType
    subject: Optional[str]
    content: str
    recipient_phone: Optional[str]
    recipient_email: Optional[str]
    status: NotificationStatus
    external_id: Optional[str]
    error_message: Optional[str]
    created_at: str
    sent_at: Optional[str]
    read_at: Optional[str]
    
    class Config:
        from_attributes = True

class MarkAsReadRequest(BaseModel):
    notification_id: int

@router.get("/", response_model=List[NotificationResponse])
async def get_user_notifications(
    unread_only: bool = False,
    limit: int = 50,
    current_user: User = Depends(get_verified_user),
    db: Session = Depends(get_db)
):
    """Get notifications for current user"""
    notifications = notification_service.get_user_notifications(
        db, current_user.id, limit, unread_only
    )
    return notifications

@router.post("/mark-read")
async def mark_notification_as_read(
    request: MarkAsReadRequest,
    current_user: User = Depends(get_verified_user),
    db: Session = Depends(get_db)
):
    """Mark notification as read"""
    success = notification_service.mark_as_read(
        db, request.notification_id, current_user.id
    )
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    return {"message": "Notification marked as read"}

@router.get("/unread-count")
async def get_unread_count(
    current_user: User = Depends(get_verified_user),
    db: Session = Depends(get_db)
):
    """Get count of unread notifications"""
    count = db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.read_at.is_(None)
    ).count()
    
    return {"unread_count": count}

# Operator endpoints
async def require_operator(current_user: User = Depends(get_verified_user)):
    if current_user.role != UserRole.OPERATOR:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Operator access required"
        )
    return current_user

class SendNotificationRequest(BaseModel):
    case_id: int
    type: NotificationType
    template: NotificationTemplateType
    variables: Optional[dict] = None

@router.post("/operator/send")
async def send_operator_notification(
    request: SendNotificationRequest,
    current_user: User = Depends(require_operator),
    db: Session = Depends(get_db)
):
    """Send notification to case client (operator only)"""
    from app.models.case import Case
    
    case = db.query(Case).filter(Case.id == request.case_id).first()
    if not case:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Case not found"
        )
    
    notification = notification_service.send_notification(
        db, case.user_id, request.type, request.template, case, request.variables
    )
    
    return {"message": "Notification sent", "notification_id": notification.id}

@router.get("/operator/recent", response_model=List[NotificationResponse])
async def get_recent_notifications(
    limit: int = 100,
    current_user: User = Depends(require_operator),
    db: Session = Depends(get_db)
):
    """Get recent notifications for all users (operator only)"""
    notifications = db.query(Notification).order_by(
        Notification.created_at.desc()
    ).limit(limit).all()
    
    return notifications

@router.get("/operator/failed", response_model=List[NotificationResponse])
async def get_failed_notifications(
    current_user: User = Depends(require_operator),
    db: Session = Depends(get_db)
):
    """Get failed notifications for troubleshooting (operator only)"""
    notifications = db.query(Notification).filter(
        Notification.status == NotificationStatus.FAILED
    ).order_by(Notification.created_at.desc()).limit(50).all()
    
    return notifications

File: frontend/lib/api/notifications.ts
import { authAPI } from './auth';

// Types for notifications
export interface Notification {
  id: number;
  user_id: number;
  case_id?: number;
  type: string;
  template: string;
  subject?: string;
  content: string;
  recipient_phone?: string;
  recipient_email?: string;
  status: string;
  external_id?: string;
  error_message?: string;
  created_at: string;
  sent_at?: string;
  read_at?: string;
}

export interface NotificationsResponse {
  notifications?: Notification[];
  error?: string;
}

export interface UnreadCountResponse {
  unread_count?: number;
  error?: string;
}

export interface MarkReadResponse {
  message?: string;
  error?: string;
}

// Notifications API client
export const notificationsApi = {
  /**
   * Get user notifications
   */
  async getUserNotifications(unreadOnly = false, limit = 50): Promise<NotificationsResponse> {
    try {
      const params = new URLSearchParams();
      if (unreadOnly) params.set('unread_only', 'true');
      if (limit) params.set('limit', limit.toString());
      
      const response = await authAPI.makeRequest(
        'GET', 
        `/notifications/?${params.toString()}`, 
        undefined, 
        true
      ) as Response;
      
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        return { 
          error: errorData.detail || `HTTP ${response.status}: ${response.statusText}` 
        };
      }

      const notifications = await response.json();
      return { notifications };
    } catch (error) {
      console.error('Get notifications error:', error);
      return { 
        error: error instanceof Error ? error.message : 'Unknown error occurred' 
      };
    }
  },

  /**
   * Get unread notifications count
   */
  async getUnreadCount(): Promise<UnreadCountResponse> {
    try {
      const response = await authAPI.makeRequest('GET', '/notifications/unread-count/', undefined, true) as Response;
      
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        return { 
          error: errorData.detail || `HTTP ${response.status}: ${response.statusText}` 
        };
      }

      const data = await response.json();
      return { unread_count: data.unread_count };
    } catch (error) {
      console.error('Get unread count error:', error);
      return { 
        error: error instanceof Error ? error.message : 'Unknown error occurred' 
      };
    }
  },

  /**
   * Mark notification as read
   */
  async markAsRead(notificationId: number): Promise<MarkReadResponse> {
    try {
      const response = await authAPI.makeRequest(
        'POST', 
        '/notifications/mark-read/', 
        { notification_id: notificationId }, 
        true
      ) as Response;
      
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        return { 
          error: errorData.detail || `HTTP ${response.status}: ${response.statusText}` 
        };
      }

      const data = await response.json();
      return { message: data.message };
    } catch (error) {
      console.error('Mark as read error:', error);
      return { 
        error: error instanceof Error ? error.message : 'Unknown error occurred' 
      };
    }
  }
};